File: word-search/word-search.cpp
class Solution {
public:
    vector<int>dir = {-1,0,1,0,-1};
    bool check(int i,int j,int n,int m)
    {
        if(i>=n || j>=m || i<0 || j<0)
        {
            return 0;
        }
        return 1;
    }
    bool bfs(vector<vector<char>>& board, int i ,int j, string& word, int k)
    {
        if(k==word.size())
        {
            return 1;
        }
        
        bool ans = 0;
        for(int x=0;x<4;x++)
        {
            if(check(i+dir[x],j+dir[x+1],board.size(),board[0].size()) && word[k]==board[i+dir[x]][j+dir[x+1]])
            {
                char curr = board[i][j];
                board[i][j]='.';
                ans = ans | bfs(board,i+dir[x],j+dir[x+1],word,k+1);
                if(ans)
                {
                    return true;
                }
                board[i][j] = curr;
            }
        }
        
        return ans;
        
    }
    bool exist(vector<vector<char>>& board, string word) 
    {
        
        int n = board.size() , m = board[0].size();
        char first = word[0];
        bool ans = 0;
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                if(board[i][j]==first)
                {
                    ans = ans | bfs(board,i,j,word,1);
                    if(ans)
                    {
                        return ans;
                    }
                }
            }
        }
        return ans; 
    }
};

File: word-search/word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        def dfs(board,i,j,word):
            
            if(len(word)==0):
                return 1
            if(i<0 or j<0 or i>=len(board) or j>=len(board[0])):
                return 0
            if(board[i][j]!=word[0]):
                return 0
            
            c=board[i][j]
            board[i][j]='.'
            
            s=word[1:len(word)]
            
            ans=0
            ans=ans | dfs(board,i+1,j,s)
            ans=ans | dfs(board,i-1,j,s)
            ans=ans | dfs(board,i,j+1,s)
            ans=ans | dfs(board,i,j-1,s)
            
            board[i][j]=c
            
            return ans
        
        for i in range(0,len(board)):
            for j in range(0,len(board[0])):
                if(board[i][j]==word[0]):
                    ans=dfs(board,i,j,word)
                    if(ans==1):
                        return 1
        return 0